// qzx-obfuscator/test-tmp/src/enums/JSFuck.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
var JSFuck;
(function (JSFuck) {
    JSFuck["False"] = "![]";
    JSFuck["True"] = "!![]";
    JSFuck["a"] = "(false+\"\")[1]";
    JSFuck["b"] = "([][\"entries\"]()+\"\")[2]";
    JSFuck["c"] = "([][\"fill\"]+\"\")[3]";
    JSFuck["d"] = "(undefined+\"\")[2]";
    JSFuck["e"] = "(true+\"\")[3]";
    JSFuck["f"] = "(false+\"\")[0]";
    JSFuck["g"] = "(false+[0]+String)[20]";
    JSFuck["h"] = "(+(101))[\"toString\"](21)[1]";
    JSFuck["i"] = "([false]+undefined)[10]";
    JSFuck["j"] = "([][\"entries\"]()+\"\")[3]";
    JSFuck["k"] = "(+(20))[\"toString\"](21)";
    JSFuck["l"] = "(false+\"\")[2]";
    JSFuck["m"] = "(Number+\"\")[11]";
    JSFuck["n"] = "(undefined+\"\")[1]";
    JSFuck["o"] = "(true+[][\"fill\"])[10]";
    JSFuck["p"] = "(+(211))[\"toString\"](31)[1]";
    JSFuck["q"] = "(+(212))[\"toString\"](31)[1]";
    JSFuck["r"] = "(true+\"\")[1]";
    JSFuck["s"] = "(false+\"\")[3]";
    JSFuck["t"] = "(true+\"\")[0]";
    JSFuck["u"] = "(undefined+\"\")[0]";
    JSFuck["v"] = "(+(31))[\"toString\"](32)";
    JSFuck["w"] = "(+(32))[\"toString\"](33)";
    JSFuck["x"] = "(+(101))[\"toString\"](34)[1]";
    JSFuck["y"] = "(NaN+[Infinity])[10]";
    JSFuck["z"] = "(+(35))[\"toString\"](36)";
    JSFuck["A"] = "(+[]+Array)[10]";
    JSFuck["B"] = "(+[]+Boolean)[10]";
    JSFuck["C"] = "Function(\"return escape\")()((\"\")[\"italics\"]())[2]";
    JSFuck["D"] = "Function(\"return escape\")()([][\"fill\"])[\"slice\"](\"-1\")";
    JSFuck["E"] = "(RegExp+\"\")[12]";
    JSFuck["F"] = "(+[]+Function)[10]";
    JSFuck["G"] = "(false+Function(\"return Date\")()())[30]";
    JSFuck["H"] = "'H'";
    JSFuck["I"] = "(Infinity+\"\")[0]";
    JSFuck["J"] = "'J'";
    JSFuck["K"] = "'K'";
    JSFuck["L"] = "'L'";
    JSFuck["M"] = "(true+Function(\"return Date\")()())[30]";
    JSFuck["N"] = "(NaN+\"\")[0]";
    JSFuck["O"] = "(NaN+Function(\"return{}\")())[11]";
    JSFuck["P"] = "'P'";
    JSFuck["Q"] = "'Q'";
    JSFuck["R"] = "(+[]+RegExp)[10]";
    JSFuck["S"] = "(+[]+String)[10]";
    JSFuck["T"] = "(NaN+Function(\"return Date\")()())[30]";
    JSFuck["U"] = "(NaN+Function(\"return{}\")()[\"toString\"][\"call\"]())[11]";
    JSFuck["V"] = "'V'";
    JSFuck["W"] = "'W'";
    JSFuck["X"] = "'X'";
    JSFuck["Y"] = "'Y'";
    JSFuck["Z"] = "'Z'";
})(JSFuck = exports.JSFuck || (exports.JSFuck = {}));
