// qzx-obfuscator/test-tmp/test/functional-tests/templates/custom-nodes/domain-lock-nodes/DomainLockNodeTemplate.spec.js
"use strict";

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; })();

Object.defineProperty(exports, "__esModule", { value: true });
var format = require("string-template");
var chai_1 = require("chai");
var ServiceIdentifiers_1 = require("../../../../../src/container/ServiceIdentifiers");
var DomainLockNodeTemplate_1 = require("../../../../../src/templates/custom-nodes/domain-lock-nodes/domain-lock-node/DomainLockNodeTemplate");
var InversifyContainerFacade_1 = require("../../../../../src/container/InversifyContainerFacade");
function getFunctionFromTemplate(templateData, callsControllerFunctionName, currentDomain) {
    var domainLockTemplate = format(DomainLockNodeTemplate_1.DomainLockNodeTemplate(), templateData);
    return Function("\n        document = {\n            domain: '" + currentDomain + "'\n        };\n\n        var " + callsControllerFunctionName + " = (function(){            \n            return function (context, fn){\t\n                return function () {\n                    return fn.apply(context, arguments);\n                };\n            }\n        })();\n\n        " + domainLockTemplate + "\n    ")();
}
describe('DomainLockNodeTemplate (): string', function () {
    var singleNodeCallControllerFunctionName = 'callsController';
    var cryptUtils = void 0;
    before(function () {
        var inversifyContainerFacade = new InversifyContainerFacade_1.InversifyContainerFacade();
        inversifyContainerFacade.load('', {});
        cryptUtils = inversifyContainerFacade.get(ServiceIdentifiers_1.ServiceIdentifiers.ICryptUtils);
    });
    describe('variant #1: current domain matches with `domainsString`', function () {
        var domainsString = ['www.example.com'].join(';');
        var currentDomain = 'www.example.com';
        var testFunc = void 0;
        before(function () {
            var _cryptUtils$hideStrin = cryptUtils.hideString(domainsString, domainsString.length * 3),
                _cryptUtils$hideStrin2 = _slicedToArray(_cryptUtils$hideStrin, 2),
                hiddenDomainsString = _cryptUtils$hideStrin2[0],
                diff = _cryptUtils$hideStrin2[1];

            testFunc = function testFunc() {
                return getFunctionFromTemplate({
                    domainLockFunctionName: 'domainLockFunction',
                    diff: diff,
                    domains: hiddenDomainsString,
                    singleNodeCallControllerFunctionName: singleNodeCallControllerFunctionName
                }, singleNodeCallControllerFunctionName, currentDomain);
            };
        });
        it('should correctly run code inside template', function () {
            chai_1.assert.doesNotThrow(testFunc);
        });
    });
    describe('variant #2: urrent domain matches with base domain of `domainsString` item', function () {
        var domainsString = ['www.test.com', '.example.com'].join(';');
        var currentDomain = 'subdomain.example.com';
        var testFunc = void 0;
        before(function () {
            var _cryptUtils$hideStrin3 = cryptUtils.hideString(domainsString, domainsString.length * 3),
                _cryptUtils$hideStrin4 = _slicedToArray(_cryptUtils$hideStrin3, 2),
                hiddenDomainsString = _cryptUtils$hideStrin4[0],
                diff = _cryptUtils$hideStrin4[1];

            testFunc = function testFunc() {
                return getFunctionFromTemplate({
                    domainLockFunctionName: 'domainLockFunction',
                    diff: diff,
                    domains: hiddenDomainsString,
                    singleNodeCallControllerFunctionName: singleNodeCallControllerFunctionName
                }, singleNodeCallControllerFunctionName, currentDomain);
            };
        });
        it('should correctly run code inside template', function () {
            chai_1.assert.doesNotThrow(testFunc);
        });
    });
    describe('variant #3: current domain doesn\'t match with `domainsString`', function () {
        var domainsString = ['www.example.com'].join(';');
        var currentDomain = 'www.test.com';
        var testFunc = void 0;
        before(function () {
            var _cryptUtils$hideStrin5 = cryptUtils.hideString(domainsString, domainsString.length * 3),
                _cryptUtils$hideStrin6 = _slicedToArray(_cryptUtils$hideStrin5, 2),
                hiddenDomainsString = _cryptUtils$hideStrin6[0],
                diff = _cryptUtils$hideStrin6[1];

            testFunc = function testFunc() {
                return getFunctionFromTemplate({
                    domainLockFunctionName: 'domainLockFunction',
                    diff: diff,
                    domains: hiddenDomainsString,
                    singleNodeCallControllerFunctionName: singleNodeCallControllerFunctionName
                }, singleNodeCallControllerFunctionName, currentDomain);
            };
        });
        it('should throw an error', function () {
            chai_1.assert.throws(testFunc);
        });
    });
});
