// qzx-obfuscator/test-tmp/test/functional-tests/javascript-obfuscator/JavaScriptObfuscator.spec.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var JavaScriptObfuscator_1 = require("../../../src/JavaScriptObfuscator");
var NoCustomNodes_1 = require("../../../src/options/presets/NoCustomNodes");
var buildLargeCode_1 = require("../../helpers/buildLargeCode");
var getRegExpMatch_1 = require("../../helpers/getRegExpMatch");
var readFileAsString_1 = require("../../helpers/readFileAsString");
describe('JavaScriptObfuscator', function () {
    describe('obfuscate (sourceCode: string, customOptions?: IObfuscatorOptions): IObfuscationResult', function () {
        describe('correct source code', function () {
            var obfuscatedCode = void 0,
                sourceMap = void 0;
            beforeEach(function () {
                var code = readFileAsString_1.readFileAsString(__dirname + '/fixtures/simple-input-1.js');
                var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, Object.assign({}, NoCustomNodes_1.NO_CUSTOM_NODES_PRESET));
                obfuscatedCode = obfuscationResult.getObfuscatedCode();
                sourceMap = obfuscationResult.getSourceMap();
            });
            it('should return correct obfuscated code', function () {
                chai_1.assert.isOk(obfuscatedCode);
            });
            it('should return empty source map', function () {
                chai_1.assert.isNotOk(sourceMap);
            });
        });
        describe('empty source code', function () {
            var obfuscatedCode = void 0;
            beforeEach(function () {
                var code = readFileAsString_1.readFileAsString(__dirname + '/fixtures/empty-input.js');
                var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code);
                obfuscatedCode = obfuscationResult.getObfuscatedCode();
            });
            it('should return an empty obfuscated code', function () {
                chai_1.assert.isNotOk(obfuscatedCode);
            });
        });
        describe('empty source code with comments', function () {
            var obfuscatedCode = void 0;
            beforeEach(function () {
                var code = readFileAsString_1.readFileAsString(__dirname + '/fixtures/comments-only.js');
                var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code);
                obfuscatedCode = obfuscationResult.getObfuscatedCode();
            });
            it('should return an empty obfuscated code', function () {
                chai_1.assert.isNotOk(obfuscatedCode);
            });
        });
        describe('`sourceMap` option is `true`', function () {
            describe('`sourceMapMode` is `separate`', function () {
                var obfuscatedCode = void 0,
                    sourceMap = void 0;
                beforeEach(function () {
                    var code = readFileAsString_1.readFileAsString(__dirname + '/fixtures/simple-input-1.js');
                    var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, Object.assign({}, NoCustomNodes_1.NO_CUSTOM_NODES_PRESET, { sourceMap: true }));
                    obfuscatedCode = obfuscationResult.getObfuscatedCode();
                    sourceMap = JSON.parse(obfuscationResult.getSourceMap()).mappings;
                });
                it('should return correct obfuscated code', function () {
                    chai_1.assert.isOk(obfuscatedCode);
                });
                it('should return correct source map', function () {
                    chai_1.assert.isOk(sourceMap);
                });
            });
            describe('`sourceMapMode` is `inline`', function () {
                var regExp = /sourceMappingURL=data:application\/json;base64/;
                var obfuscatedCode = void 0,
                    sourceMap = void 0;
                beforeEach(function () {
                    var code = readFileAsString_1.readFileAsString(__dirname + '/fixtures/simple-input-1.js');
                    var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, Object.assign({}, NoCustomNodes_1.NO_CUSTOM_NODES_PRESET, { sourceMap: true, sourceMapMode: 'inline' }));
                    obfuscatedCode = obfuscationResult.getObfuscatedCode();
                    sourceMap = JSON.parse(obfuscationResult.getSourceMap()).mappings;
                });
                it('should return correct obfuscated code', function () {
                    chai_1.assert.isOk(obfuscatedCode);
                });
                it('should return obfuscated code with inline source map as Base64 string', function () {
                    chai_1.assert.match(obfuscatedCode, regExp);
                });
                it('should return correct source map', function () {
                    chai_1.assert.isOk(sourceMap);
                });
            });
            describe('empty source code', function () {
                var obfuscatedCode = void 0,
                    sourceMapNames = void 0,
                    sourceMapSources = void 0,
                    sourceMapMappings = void 0;
                beforeEach(function () {
                    var code = readFileAsString_1.readFileAsString(__dirname + '/fixtures/empty-input.js');
                    var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, {
                        sourceMap: true
                    });
                    obfuscatedCode = obfuscationResult.getObfuscatedCode();
                    var sourceMapObject = JSON.parse(obfuscationResult.getSourceMap());
                    sourceMapNames = sourceMapObject.names;
                    sourceMapSources = sourceMapObject.sources;
                    sourceMapMappings = sourceMapObject.mappings;
                });
                it('should return empty obfuscated code', function () {
                    chai_1.assert.isNotOk(obfuscatedCode);
                });
                it('should return empty source map property `names`', function () {
                    chai_1.assert.deepEqual(sourceMapNames, []);
                });
                it('should return empty source map property `sources`', function () {
                    chai_1.assert.deepEqual(sourceMapSources, []);
                });
                it('should return empty source map property `mappings`', function () {
                    chai_1.assert.isNotOk(sourceMapMappings);
                });
            });
        });
        describe('variable inside global scope', function () {
            var regExp = /^var *test *= *0x\d+;$/;
            var obfuscatedCode = void 0;
            beforeEach(function () {
                var code = readFileAsString_1.readFileAsString(__dirname + '/fixtures/simple-input-1.js');
                var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, Object.assign({}, NoCustomNodes_1.NO_CUSTOM_NODES_PRESET));
                obfuscatedCode = obfuscationResult.getObfuscatedCode();
            });
            it('should return correct obfuscated code', function () {
                chai_1.assert.match(obfuscatedCode, regExp);
            });
        });
        describe('variable inside global scope', function () {
            var regExp = /^\(function *\(\) *\{ *var *_0x[\w]+ *= *0x\d+; *\}(\(\)\)|\)\(\));?$/;
            var obfuscatedCode = void 0;
            beforeEach(function () {
                var code = readFileAsString_1.readFileAsString(__dirname + '/fixtures/block-scope.js');
                var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, Object.assign({}, NoCustomNodes_1.NO_CUSTOM_NODES_PRESET));
                obfuscatedCode = obfuscationResult.getObfuscatedCode();
            });
            it('should return correct obfuscated code', function () {
                chai_1.assert.match(obfuscatedCode, regExp);
            });
        });
        describe('latin literal variable value', function () {
            var stringArrayLatinRegExp = /^var _0x(\w){4} *= *\['abc'\];/;
            var stringArrayCallRegExp = /var *test *= *_0x(\w){4}\('0x0'\);$/;
            var obfuscatedCode = void 0;
            beforeEach(function () {
                var code = readFileAsString_1.readFileAsString(__dirname + '/fixtures/simple-input-2.js');
                var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, Object.assign({}, NoCustomNodes_1.NO_CUSTOM_NODES_PRESET, { stringArray: true, stringArrayThreshold: 1 }));
                obfuscatedCode = obfuscationResult.getObfuscatedCode();
            });
            it('match #1: should return correct obfuscated code', function () {
                chai_1.assert.match(obfuscatedCode, stringArrayLatinRegExp);
            });
            it('match #2: should return correct obfuscated code', function () {
                chai_1.assert.match(obfuscatedCode, stringArrayCallRegExp);
            });
        });
        describe('cyrillic literal variable value', function () {
            var stringArrayCyrillicRegExp = /^var _0x(\w){4} *= *\['абц'\];/;
            var stringArrayCallRegExp = /var *test *= *_0x(\w){4}\('0x0'\);$/;
            var obfuscatedCode = void 0;
            beforeEach(function () {
                var code = readFileAsString_1.readFileAsString(__dirname + '/fixtures/simple-input-cyrillic.js');
                var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, Object.assign({}, NoCustomNodes_1.NO_CUSTOM_NODES_PRESET, { stringArray: true, stringArrayThreshold: 1 }));
                obfuscatedCode = obfuscationResult.getObfuscatedCode();
            });
            it('match #1: should return correct obfuscated code', function () {
                chai_1.assert.match(obfuscatedCode, stringArrayCyrillicRegExp);
            });
            it('match #2: should return correct obfuscated code', function () {
                chai_1.assert.match(obfuscatedCode, stringArrayCallRegExp);
            });
        });
        describe('seed', function () {
            this.timeout(60000);
            describe('same seed on each run', function () {
                var code = readFileAsString_1.readFileAsString('./test/fixtures/sample.js');
                var samples = 100;
                var obfuscatedCode1 = void 0,
                    obfuscatedCode2 = void 0,
                    seed = 12345,
                    equalsCount = 0;
                beforeEach(function () {
                    for (var i = 0; i < samples; i++) {
                        if (i % 20 === 0) {
                            seed++;
                        }
                        var obfuscationResult1 = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, {
                            seed: seed
                        });
                        var obfuscationResult2 = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, {
                            seed: seed
                        });
                        obfuscatedCode1 = obfuscationResult1.getObfuscatedCode();
                        obfuscatedCode2 = obfuscationResult2.getObfuscatedCode();
                        if (obfuscatedCode1 === obfuscatedCode2) {
                            equalsCount++;
                        }
                    }
                });
                it('should return same code every time with same `seed`', function () {
                    chai_1.assert.equal(equalsCount, samples);
                });
            });
            describe('variant #1: different seed on each run', function () {
                var code = readFileAsString_1.readFileAsString('./test/fixtures/sample.js');
                var obfuscatedCode1 = void 0,
                    obfuscatedCode2 = void 0;
                beforeEach(function () {
                    var obfuscationResult1 = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, {
                        seed: 12345
                    });
                    var obfuscationResult2 = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, {
                        seed: 12346
                    });
                    obfuscatedCode1 = obfuscationResult1.getObfuscatedCode();
                    obfuscatedCode2 = obfuscationResult2.getObfuscatedCode();
                });
                it('should return different obfuscated code with different `seed` option value', function () {
                    chai_1.assert.notEqual(obfuscatedCode1, obfuscatedCode2);
                });
            });
            describe('variant #2: different seed on each run', function () {
                var code = readFileAsString_1.readFileAsString('./test/fixtures/sample.js');
                var obfuscatedCode1 = void 0,
                    obfuscatedCode2 = void 0;
                beforeEach(function () {
                    var obfuscationResult1 = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, {
                        seed: 0
                    });
                    var obfuscationResult2 = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, {
                        seed: 0
                    });
                    obfuscatedCode1 = obfuscationResult1.getObfuscatedCode();
                    obfuscatedCode2 = obfuscationResult2.getObfuscatedCode();
                });
                it('should return different obfuscated code with different `seed` option value', function () {
                    chai_1.assert.notEqual(obfuscatedCode1, obfuscatedCode2);
                });
            });
            describe('variant #3: same seed for different source code', function () {
                var code1 = readFileAsString_1.readFileAsString(__dirname + '/fixtures/simple-input-cyrillic.js');
                var code2 = readFileAsString_1.readFileAsString(__dirname + '/fixtures/simple-input-2.js');
                var regExp = /var (_0x(\w){4}) *= *\['.*'\];/;
                var match1 = void 0,
                    match2 = void 0;
                beforeEach(function () {
                    var obfuscationResult1 = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code1, {
                        seed: 123
                    });
                    var obfuscationResult2 = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code2, {
                        seed: 123
                    });
                    var obfuscatedCode1 = obfuscationResult1.getObfuscatedCode();
                    var obfuscatedCode2 = obfuscationResult2.getObfuscatedCode();
                    match1 = getRegExpMatch_1.getRegExpMatch(obfuscatedCode1, regExp);
                    match2 = getRegExpMatch_1.getRegExpMatch(obfuscatedCode2, regExp);
                });
                it('should return different String Array names for different source code with same seed', function () {
                    chai_1.assert.notEqual(match1, match2);
                });
            });
        });
        describe('new.target MetaProperty', function () {
            var regExp = /new\.target *=== *Foo/;
            var obfuscatedCode = void 0;
            beforeEach(function () {
                var code = readFileAsString_1.readFileAsString(__dirname + '/fixtures/new-target.js');
                var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, Object.assign({}, NoCustomNodes_1.NO_CUSTOM_NODES_PRESET));
                obfuscatedCode = obfuscationResult.getObfuscatedCode();
            });
            it('should keep new.target MetaProperty', function () {
                chai_1.assert.match(obfuscatedCode, regExp);
            });
        });
        describe('mangle', function () {
            var regExp = /var *a *= *0x1/;
            var obfuscatedCode = void 0;
            beforeEach(function () {
                var code = readFileAsString_1.readFileAsString(__dirname + '/fixtures/mangle.js');
                var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, {
                    mangle: true
                });
                obfuscatedCode = obfuscationResult.getObfuscatedCode();
            });
            it('should mangle obfuscated code', function () {
                chai_1.assert.match(obfuscatedCode, regExp);
            });
        });
        describe('3.5k variables', function () {
            this.timeout(200000);
            var expectedValue = 3500;
            var result = void 0;
            beforeEach(function () {
                var code = buildLargeCode_1.buildLargeCode(expectedValue);
                var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, {
                    compact: true,
                    controlFlowFlattening: true,
                    controlFlowFlatteningThreshold: 1,
                    deadCodeInjection: true,
                    deadCodeInjectionThreshold: 1,
                    disableConsoleOutput: false,
                    rotateStringArray: true,
                    stringArray: true,
                    stringArrayEncoding: 'rc4',
                    stringArrayThreshold: 1,
                    unicodeEscapeSequence: false
                });
                var obfuscatedCode = obfuscationResult.getObfuscatedCode();
                result = eval(obfuscatedCode);
            });
            it('should correctly obfuscate 3.5k variables', function () {
                chai_1.assert.equal(result, expectedValue);
            });
        });
    });
});
