// qzx-obfuscator/test-tmp/test/unit-tests/decorators/initializable/Initializable.spec.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var chai_1 = require("chai");
var Initializable_1 = require("../../../../src/decorators/Initializable");
describe('@initializable', function () {
    describe('initializable (initializeMethodKey: string): any', function () {
        describe('variant #1: property was initialized', function () {
            var testFunc = function testFunc() {
                var Foo = function () {
                    function Foo() {
                        _classCallCheck(this, Foo);
                    }

                    _createClass(Foo, [{
                        key: "initialize",
                        value: function initialize(property) {
                            this.property = property;
                        }
                    }]);

                    return Foo;
                }();

                tslib_1.__decorate([Initializable_1.initializable(), tslib_1.__metadata("design:type", String)], Foo.prototype, "property", void 0);
                var foo = new Foo();
                foo.initialize('baz');
                foo.property;
            };
            it('shouldn\'t throws an errors if property was initialized', function () {
                chai_1.assert.doesNotThrow(testFunc, Error);
            });
        });
        describe('variant #2: custom initialization method name is passed', function () {
            var testFunc = function testFunc() {
                var Foo = function () {
                    function Foo() {
                        _classCallCheck(this, Foo);
                    }

                    _createClass(Foo, [{
                        key: "initialize",
                        value: function initialize() {}
                    }, {
                        key: "bar",
                        value: function bar(property) {
                            this.property = property;
                        }
                    }]);

                    return Foo;
                }();

                tslib_1.__decorate([Initializable_1.initializable(), tslib_1.__metadata("design:type", String)], Foo.prototype, "property", void 0);
                var foo = new Foo();
                foo.bar('baz');
                foo.property;
            };
            it('shouldn\'t throws an errors if custom initialization method name is passed', function () {
                chai_1.assert.doesNotThrow(testFunc, Error);
            });
        });
        describe('variant #3: property didn\'t initialized', function () {
            var testFunc = function testFunc() {
                var Foo = function () {
                    function Foo() {
                        _classCallCheck(this, Foo);
                    }

                    _createClass(Foo, [{
                        key: "initialize",
                        value: function initialize(property) {}
                    }]);

                    return Foo;
                }();

                tslib_1.__decorate([Initializable_1.initializable(), tslib_1.__metadata("design:type", String)], Foo.prototype, "property", void 0);
                var foo = new Foo();
                foo.initialize('baz');
                foo.property;
            };
            it('should throws an error if property didn\'t initialized', function () {
                chai_1.assert.throws(testFunc, /Property `property` is not initialized/);
            });
        });
        describe('variant #4: `initialize` method with custom name', function () {
            var testFunc = function testFunc() {
                var Foo = function () {
                    function Foo() {
                        _classCallCheck(this, Foo);
                    }

                    _createClass(Foo, [{
                        key: "initialize",
                        value: function initialize(property) {
                            this.property = property;
                        }
                    }]);

                    return Foo;
                }();

                tslib_1.__decorate([Initializable_1.initializable('bar'), tslib_1.__metadata("design:type", String)], Foo.prototype, "property", void 0);
                var foo = new Foo();
                foo.initialize('baz');
                foo.property;
            };
            it('should throws an error if `initialize` method with custom name wasn\'t found', function () {
                chai_1.assert.throws(testFunc, /method with initialization logic not found/);
            });
        });
    });
});
