// qzx-obfuscator/test-tmp/src/enums/container/custom-nodes/ControlFlowCustomNode.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
var ControlFlowCustomNode;
(function (ControlFlowCustomNode) {
    ControlFlowCustomNode[ControlFlowCustomNode["BinaryExpressionFunctionNode"] = 0] = "BinaryExpressionFunctionNode";
    ControlFlowCustomNode[ControlFlowCustomNode["BlockStatementControlFlowFlatteningNode"] = 1] = "BlockStatementControlFlowFlatteningNode";
    ControlFlowCustomNode[ControlFlowCustomNode["CallExpressionControlFlowStorageCallNode"] = 2] = "CallExpressionControlFlowStorageCallNode";
    ControlFlowCustomNode[ControlFlowCustomNode["CallExpressionFunctionNode"] = 3] = "CallExpressionFunctionNode";
    ControlFlowCustomNode[ControlFlowCustomNode["ControlFlowStorageNode"] = 4] = "ControlFlowStorageNode";
    ControlFlowCustomNode[ControlFlowCustomNode["ExpressionWithOperatorControlFlowStorageCallNode"] = 5] = "ExpressionWithOperatorControlFlowStorageCallNode";
    ControlFlowCustomNode[ControlFlowCustomNode["LogicalExpressionFunctionNode"] = 6] = "LogicalExpressionFunctionNode";
    ControlFlowCustomNode[ControlFlowCustomNode["StringLiteralControlFlowStorageCallNode"] = 7] = "StringLiteralControlFlowStorageCallNode";
    ControlFlowCustomNode[ControlFlowCustomNode["StringLiteralNode"] = 8] = "StringLiteralNode";
})(ControlFlowCustomNode = exports.ControlFlowCustomNode || (exports.ControlFlowCustomNode = {}));
