// qzx-obfuscator/test-tmp/src/enums/container/custom-nodes/CustomNode.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
var CustomNode;
(function (CustomNode) {
    CustomNode[CustomNode["ConsoleOutputDisableExpressionNode"] = 0] = "ConsoleOutputDisableExpressionNode";
    CustomNode[CustomNode["DebugProtectionFunctionCallNode"] = 1] = "DebugProtectionFunctionCallNode";
    CustomNode[CustomNode["DebugProtectionFunctionIntervalNode"] = 2] = "DebugProtectionFunctionIntervalNode";
    CustomNode[CustomNode["DebugProtectionFunctionNode"] = 3] = "DebugProtectionFunctionNode";
    CustomNode[CustomNode["DomainLockNode"] = 4] = "DomainLockNode";
    CustomNode[CustomNode["NodeCallsControllerFunctionNode"] = 5] = "NodeCallsControllerFunctionNode";
    CustomNode[CustomNode["SelfDefendingUnicodeNode"] = 6] = "SelfDefendingUnicodeNode";
    CustomNode[CustomNode["StringArrayCallsWrapper"] = 7] = "StringArrayCallsWrapper";
    CustomNode[CustomNode["StringArrayNode"] = 8] = "StringArrayNode";
    CustomNode[CustomNode["StringArrayRotateFunctionNode"] = 9] = "StringArrayRotateFunctionNode";
})(CustomNode = exports.CustomNode || (exports.CustomNode = {}));
