// qzx-obfuscator/test-tmp/src/enums/container/node-transformers/NodeTransformer.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
var NodeTransformer;
(function (NodeTransformer) {
    NodeTransformer[NodeTransformer["BlockStatementControlFlowTransformer"] = 0] = "BlockStatementControlFlowTransformer";
    NodeTransformer[NodeTransformer["DeadCodeInjectionTransformer"] = 1] = "DeadCodeInjectionTransformer";
    NodeTransformer[NodeTransformer["FunctionControlFlowTransformer"] = 2] = "FunctionControlFlowTransformer";
    NodeTransformer[NodeTransformer["CatchClauseTransformer"] = 3] = "CatchClauseTransformer";
    NodeTransformer[NodeTransformer["FunctionDeclarationTransformer"] = 4] = "FunctionDeclarationTransformer";
    NodeTransformer[NodeTransformer["FunctionTransformer"] = 5] = "FunctionTransformer";
    NodeTransformer[NodeTransformer["LabeledStatementTransformer"] = 6] = "LabeledStatementTransformer";
    NodeTransformer[NodeTransformer["LiteralTransformer"] = 7] = "LiteralTransformer";
    NodeTransformer[NodeTransformer["MemberExpressionTransformer"] = 8] = "MemberExpressionTransformer";
    NodeTransformer[NodeTransformer["MethodDefinitionTransformer"] = 9] = "MethodDefinitionTransformer";
    NodeTransformer[NodeTransformer["ObjectExpressionTransformer"] = 10] = "ObjectExpressionTransformer";
    NodeTransformer[NodeTransformer["TemplateLiteralTransformer"] = 11] = "TemplateLiteralTransformer";
    NodeTransformer[NodeTransformer["VariableDeclarationTransformer"] = 12] = "VariableDeclarationTransformer";
})(NodeTransformer = exports.NodeTransformer || (exports.NodeTransformer = {}));
