// qzx-obfuscator/test-tmp/src/container/InversifyContainerFacade.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var ServiceIdentifiers_1 = require("./ServiceIdentifiers");
var ControlFlowTransformersModule_1 = require("./modules/node-transformers/ControlFlowTransformersModule");
var CustomNodesModule_1 = require("./modules/custom-nodes/CustomNodesModule");
var ObfuscatingTransformersModule_1 = require("./modules/node-transformers/ObfuscatingTransformersModule");
var NodeTransformersModule_1 = require("./modules/node-transformers/NodeTransformersModule");
var StackTraceAnalyzerModule_1 = require("./modules/stack-trace-analyzer/StackTraceAnalyzerModule");
var StoragesModule_1 = require("./modules/storages/StoragesModule");
var UtilsModule_1 = require("./modules/utils/UtilsModule");
var JavaScriptObfuscatorInternal_1 = require("../JavaScriptObfuscatorInternal");
var ObfuscationEventEmitter_1 = require("../event-emitters/ObfuscationEventEmitter");
var ObfuscationResult_1 = require("../ObfuscationResult");
var Obfuscator_1 = require("../Obfuscator");
var Options_1 = require("../options/Options");
var SourceCode_1 = require("../SourceCode");
var SourceMapCorrector_1 = require("../SourceMapCorrector");

var InversifyContainerFacade = function () {
    function InversifyContainerFacade() {
        _classCallCheck(this, InversifyContainerFacade);

        this.container = new inversify_1.Container();
    }

    _createClass(InversifyContainerFacade, [{
        key: "get",
        value: function get(serviceIdentifier) {
            return this.container.get(serviceIdentifier);
        }
    }, {
        key: "getNamed",
        value: function getNamed(serviceIdentifier, named) {
            return this.container.getNamed(serviceIdentifier, named);
        }
    }, {
        key: "load",
        value: function load(sourceCode, options) {
            this.container.bind(ServiceIdentifiers_1.ServiceIdentifiers.ISourceCode).toDynamicValue(function () {
                return new SourceCode_1.SourceCode(sourceCode);
            }).inSingletonScope();
            this.container.bind(ServiceIdentifiers_1.ServiceIdentifiers.IOptions).toDynamicValue(function () {
                return new Options_1.Options(options);
            }).inSingletonScope();
            this.container.bind(ServiceIdentifiers_1.ServiceIdentifiers.IJavaScriptObfuscator).to(JavaScriptObfuscatorInternal_1.JavaScriptObfuscatorInternal).inSingletonScope();
            this.container.bind(ServiceIdentifiers_1.ServiceIdentifiers.IObfuscator).to(Obfuscator_1.Obfuscator).inSingletonScope();
            this.container.bind(ServiceIdentifiers_1.ServiceIdentifiers.IObfuscationResult).to(ObfuscationResult_1.ObfuscationResult).inSingletonScope();
            this.container.bind(ServiceIdentifiers_1.ServiceIdentifiers.Factory__IObfuscationResult).toFactory(function (context) {
                return function (obfuscatedCode, sourceMap) {
                    var obfuscationResult = context.container.get(ServiceIdentifiers_1.ServiceIdentifiers.IObfuscationResult);
                    obfuscationResult.initialize(obfuscatedCode, sourceMap);
                    return obfuscationResult;
                };
            });
            this.container.bind(ServiceIdentifiers_1.ServiceIdentifiers.ISourceMapCorrector).to(SourceMapCorrector_1.SourceMapCorrector).inSingletonScope();
            this.container.bind(ServiceIdentifiers_1.ServiceIdentifiers.IObfuscationEventEmitter).to(ObfuscationEventEmitter_1.ObfuscationEventEmitter).inSingletonScope();
            this.container.load(UtilsModule_1.utilsModule);
            this.container.load(StoragesModule_1.storagesModule);
            this.container.load(StackTraceAnalyzerModule_1.stackTraceAnalyzerModule);
            this.container.load(CustomNodesModule_1.customNodesModule);
            this.container.load(NodeTransformersModule_1.nodeTransformersModule);
            this.container.load(ControlFlowTransformersModule_1.controlFlowTransformersModule);
            this.container.load(ObfuscatingTransformersModule_1.obfuscatingTransformersModule);
        }
    }, {
        key: "unload",
        value: function unload() {
            this.container.unbindAll();
        }
    }], [{
        key: "getFactory",
        value: function getFactory(serviceIdentifier) {
            return function (context) {
                return function (bindingName) {
                    return context.container.getNamed(serviceIdentifier, bindingName);
                };
            };
        }
    }, {
        key: "getCacheFactory",
        value: function getCacheFactory(serviceIdentifier) {
            return function (context) {
                var cache = new Map();
                return function (bindingName) {
                    if (cache.has(bindingName)) {
                        return cache.get(bindingName);
                    }
                    var object = context.container.getNamed(serviceIdentifier, bindingName);
                    cache.set(bindingName, object);
                    return object;
                };
            };
        }
    }, {
        key: "getConstructorFactory",
        value: function getConstructorFactory(serviceIdentifier) {
            for (var _len = arguments.length, dependencies = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
                dependencies[_key - 1] = arguments[_key];
            }

            return function (context) {
                var cache = new Map();
                var cachedDependencies = [];
                return function (bindingName) {
                    dependencies.forEach(function (dependency, index) {
                        if (!cachedDependencies[index]) {
                            cachedDependencies[index] = context.container.get(dependency);
                        }
                    });
                    if (cache.has(bindingName)) {
                        return new (Function.prototype.bind.apply(cache.get(bindingName), [null].concat(cachedDependencies)))();
                    }
                    var constructor = context.container.getNamed(serviceIdentifier, bindingName);
                    cache.set(bindingName, constructor);
                    return new (Function.prototype.bind.apply(constructor, [null].concat(cachedDependencies)))();
                };
            };
        }
    }]);

    return InversifyContainerFacade;
}();

exports.InversifyContainerFacade = InversifyContainerFacade;
