// qzx-obfuscator/test-tmp/test/helpers/buildLargeCode.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", { value: true });
function buildLargeCode(linesOfCode) {
    return new LargeCodeBuilder(linesOfCode).build();
}
exports.buildLargeCode = buildLargeCode;

var LargeCodeBuilder = function () {
    function LargeCodeBuilder(linesOfCode) {
        _classCallCheck(this, LargeCodeBuilder);

        this.code = '';
        this.linesOfCode = linesOfCode;
    }

    _createClass(LargeCodeBuilder, [{
        key: "build",
        value: function build() {
            var lastLineIndex = this.linesOfCode - 1;
            var funcIndex = 0,
                isFuncWasOpened = false;
            this.addLine("function Foo () {");
            this.addLine("var var0 = 0;");
            for (var index = 0; index < this.linesOfCode; index++) {
                var step = index % 10;
                var newIndex = index + 1;
                var isLastLine = index === lastLineIndex;
                if (step === 3) {
                    funcIndex = index;
                    isFuncWasOpened = true;
                    this.addLine("function func" + funcIndex + " () {");
                    this.addLine("var var" + newIndex + " = var" + index + ";");
                }
                if (step === 4) {
                    this.addLine("if (true) {");
                }
                if (step === 6) {
                    this.addLine("if (true) {");
                }
                this.addLine("var var" + newIndex + " = var" + index + ";");
                this.addLine("var" + newIndex + "++;");
                if (step === 9 || isLastLine && isFuncWasOpened) {
                    isFuncWasOpened = false;
                    this.addLine("}");
                    this.addLine("}");
                    this.addLine("return var" + newIndex + ";");
                    this.addLine("}");
                    this.addLine("var var" + newIndex + " = func" + funcIndex + "();");
                }
                if (isLastLine) {
                    this.addLine("result = var" + newIndex + ";");
                }
            }
            this.addLine("return result;");
            this.addLine("}");
            this.addLine("Foo();");
            return this.code;
        }
    }, {
        key: "addLine",
        value: function addLine(line) {
            this.code += "\n" + line;
        }
    }]);

    return LargeCodeBuilder;
}();
