// qzx-obfuscator/test-tmp/test/runtime-tests/JavaScriptObfuscatorRuntime.spec.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var readFileAsString_1 = require("../helpers/readFileAsString");
var JavaScriptObfuscator_1 = require("../../src/JavaScriptObfuscator");
describe('JavaScriptObfuscator runtime eval', function () {
    this.timeout(100000);
    var obfuscatedCode = void 0;
    before(function () {
        var code = readFileAsString_1.readFileAsString(__dirname + '/fixtures/sha256.js');
        var obfuscationResult = JavaScriptObfuscator_1.JavaScriptObfuscator.obfuscate(code, {
            controlFlowFlattening: true,
            deadCodeInjection: true,
            debugProtection: true,
            selfDefending: true,
            stringArrayEncoding: 'rc4'
        });
        obfuscatedCode = obfuscationResult.getObfuscatedCode();
    });
    it('should obfuscate code without any runtime errors after obfuscation: variant #1 sha256', function () {
        chai_1.assert.equal(eval(obfuscatedCode + " sha256('test');"), '9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08');
    });
});
